blog/tanstack-router-intro/index.md

```diff
@@ -0,0 +1,461 @@
+      tags: Array.isArray(search.tags) ? search.tags : [],
+    };
+  },
+  component: ({}) => {
+```
+
+Note that (unlike `URLSearchParams`) we are not limited to just string values. We can put objects or arrays in there, and TanStack will do the work of serializing and de-serializing it for us. Not only that, but you can even specify [custom serialization mechanisms](https://tanstack.com/router/latest/docs/framework/react/guide/custom-search-param-serialization).
+
+Let's manually browse to this path, without any search params, and see what happens. When we browse to
+
+```
+http://localhost:5173/epics/1/milestones
+```
+
+Router replaces (does not redirect) us to
+
+```
+http://localhost:5173/epics/1/milestones?page=1&search=&tags=%5B%5D
+```
+
+TanStack ran our validation function, and then replaced our url with the correct, valid search params. If you don't like that it forces the url to be "ugly" like that, stay tuned; that's easily worked around. But first let's work with what we have.
+
+We've been using the `Route.useParams` method multiple times. There's also a `Route.useSearch` that does the same thing, for search params. But let's do something a little different. We've previously been putting everything in the same route file, so we could just directly reference the Route object from the same lexical scope. Let's build a separate component to read, and update these search params.
+
+I've added a `MilestoneSearch.tsx` component. You might think you could just import the `Route` object from the route file. But that's dangerous. You're likely to create a circular dependency, which might or might not work, depending on your bundler. And even if it "works" you might have some hidden issues lurking.
+
+Fortunately Router gives you a direct api to handle this: `getRouteApi` which is exported from `@tanstack/react-router`. We pass it a (statically typed) route, and it gives us back the correct route object
+
+```ts
+const route = getRouteApi("/epics/$epicId/milestones/");
+```
+
+and now we can call `useSearch` on that route object and get our statically typed result.
+
+![get search params](/tanstack-router-intro/get-search-params.jpg)
+
+We won't belabor the form elements and click handlers to sync and gather new values for these search parameters. Let's just assume we have some new values, and see how we set them. For this, we can use the `useNavigate` hook.
+
+```ts
+const navigate = useNavigate({ from: "/epics/$epicId/milestones/" });
+```
+
+We call it, and tell it where we're navigating _from_. Now we use the result and tell it where we want to _go_ (the same place we are), and are given a `search` function from which we return the new search params. Naturally, TypeScript will yell at us if we leave anything off. As a convenience, Router will pass this search function the current values, making it easy to just add / override something. So to page up, we can do
+
+```ts
+navigate({
+  to: ".",
+  search: prev => {
```

Comment:
we detect whether ALL search params of the `to` target are optional, and if yes, you can omit the search property in total.
might be worth noting this.

Reply:
also, if you happen to navigate from route A to route B and route B shares some of the _search_ params that route A has, then you only need to specify the ones that B has but A does not have.
you can of course also specify search params that A has, but if you omit those, they will be carried over from A 

this only works if you specify `from`!

almost the same is true for _path_ params.
Here we also compare the path params of `from` and `to`, but we only make those overlapping path params optional that reside in the same hiearchy

```tsx
navigate({from: '/foo/$id', to: '/foo/$id/bar})
```
this won't require the `id` path param to be specified

however 

```tsx
navigate({from: '/foo/$id', to: '/bar/$id})
```

will required `id` to be specified, as we assume this is another kind of `id`.

and finally:

```tsx
navigate({from: '/foo/$id', to: '/foo/$id/bar/$search, params: {search: 'asdf'} })
```

`id` is optional, but `search` is required here

Reply:
and: `navigate` and `<Link>` share the same API, so all of the above is true for `<Link>` as well

Reply:
Super good notes - extremely useful for me - thank you! That said, I don't think I have any examples rich enough to fit that in